### PR TITLE
[BOUNTY] Liquids try #3: Return of the pee

### PR DESCRIPTION
**Pull request by ToasterBiome**
# Document the changes in your pull request

Ports some of https://github.com/Monkestation/Monkestation2.0/commit/8bbcd47bcbef614311a5a1a2819c25fff16e8709
Ports https://github.com/Monkestation/Monkestation2.0/pull/2072
Ports https://github.com/tgstation/tgstation/pull/71430

fulfills https://forums.yogstation.net/threads/300-liquid-simulation-tgcode-edition-ss13-server.27154/

Adds liquids from Monkestation. Adds some basic ocean turfs.

Things this does have:
Liquids
Interaction with liquids
Admin ability to add/remove liquids
Drains?

Things this does not have:
TG plumbing (further PR)
Ocean generation (further PR)

# Why is this good for the game?

More stuff to do and interact with = more fun

# Testing
dreamseeker_z2PxOVK3J2
dreamseeker_o7TSK5nLuC

# Wiki Documentation

will fill out

# Changelog

:cl:  
rscadd: adds liquids
/:cl:

### CODE REVIEW

**Comment by neeshacark** [comment]
finally, it is our time

**Comment by cowbot92** [comment]
im gonna pee so hard

**Comment by AMyriad** [comment]
one step closer to ocean station inshallah

**Comment by wonderinghost** [comment]
there better be jarate gameplay

**Comment by Moltijoe** [comment]
Is there any way to tell a difference between a liquid and gas at a glance? They both just seem to be translucent overlays

**Comment by neeshacark** [comment]
> Is there any way to tell a difference between a liquid and gas at a glance? They both just seem to be translucent overlays

its pretty obvious

**Comment by Chubbygummibear** [comment]
check out how hard i can

**Comment by ShoziOffical** [comment]
> check out how hard i can

bro got a third stick

**Comment by Chubbygummibear** [comment]
> > check out how hard i can
> 
> bro got a third stick

it's called a dixel

**Comment by Beeper-Skeeper** [comment]
PISS

**Comment by Beeper-Skeeper** [comment]
now we need Jarate

File: yogstation/code/modules/liquids/height_floors.dm [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+/obj/item/stack/tile/lowered/iron
+	name = "lowered floor tile"
+	singular_name = "lowered floor tile"
+	turf_type = /turf/open/floor/lowered
+	merge_type = /obj/item/stack/tile/lowered
+	icon = 'yogstation/icons/obj/items/tiles.dmi'
+	icon_state = "lowered"
+
+/turf/open/floor/iron/pool/rust_heretic_act()
```
Comment (Moltijoe) [inline yogstation/code/modules/liquids/height_floors.dm:25]: Where is this declared

**Review by cuackles — APPROVED** [review]

File: yogstation/code/modules/liquids/height_floors.dm (hunk 1 above)
Comment (ToasterBiome) [reply yogstation/code/modules/liquids/height_floors.dm:25]: > Where is this declared

right there

Comment (Moltijoe) [reply yogstation/code/modules/liquids/height_floors.dm:25]: gotcha, sure why not

**Comment by Moltijoe** [comment]
water still gets stuck in dense structures
image

space doesn't drain water
image

also, breaking welding fuel tanks and water tanks should release the respective liquid

water vapour should maybe condense into water puddles rather than making the floor wet
hell, the old "wet floor" system should be replaced with reagent puddles
image

maints should probably also get access to the water spawning admin proc

**Comment by Moltijoe** [comment]
hitting a surface with a beaker should probably just dump the liquid on the previous tile
image

**Comment by Moltijoe** [comment]
water was being blocked by a space tile and when the space tile was filled in and changed to a regular turf tile it didn't update until water was specifically applied to that tile that used to be space

image

**Comment by Moltijoe** [comment]
closing doors will delete the water on that tile
opening that door will splash water onto that tile even if the water isn't high enough
splashes should probably only happen if there's a certain level of volume difference
image

**Comment by Moltijoe** [comment]
puddles don't like merging
image

File: yogstation/code/modules/liquids/liquid_groups.dm [hunk 2]
```diff
@@ -0,0 +1,1024 @@
...
+
+	return TRUE
+
+/datum/liquid_group/proc/water_rush(turf/new_turf, turf/source_turf)
+	var/direction = get_dir(source_turf, new_turf)
+	for(var/atom/movable/target_atom in new_turf)
+		if(!target_atom.anchored && !target_atom.pulledby && !isobserver(target_atom) && (target_atom.move_resist < INFINITY))
+			reagents.reaction(target_atom, TOUCH, (reagents_per_turf * 0.5))
+			if(expected_turf_height < LIQUID_ANKLES_LEVEL_HEIGHT)
+				return
```
Comment (Moltijoe) [inline yogstation/code/modules/liquids/liquid_groups.dm]: probably worth putting reagents.reaction below the check
if the liquid isn't enough to push them, chances are it isn't gonna react with them

also, the react with reagents should probably still happen, even if the target is anchored or being pulled, or has infinity move resist

also also, magboots should probably resist this, checking heavy gravity is what it does

**Review by Moltijoe — CHANGES_REQUESTED** [review]

**Review by Moltijoe — APPROVED** [review]
fine, i guess it's alright, i can make those tweaks in a different pr
nothing fundamentally flawed with the pr

**Review by Moltijoe — APPROVED** [review]

**Review by JamieD1 — APPROVED** [review]
:/

**Pull request merged**